DeepSpeech.ipynb

```diff
@@ -1255,13 +1214,75 @@
    },
    "outputs": [],
    "source": [
-    "def calculate_and_print_wer_report(session, caption, tower_decodings, tower_labels, tower_total_losses, show_ranked=True):\n",
-    "    items, mean = calculate_wer(session, tower_decodings, tower_labels, tower_total_losses)\n",
-    "    if show_ranked:\n",
-    "        print_wer_report(caption, mean, items=items)\n",
-    "    else:\n",
-    "        print_wer_report(caption, mean)\n",
-    "    return items, mean"
+    "def collect_results(results_tuple, returns):\n",
+    "    # Each of the arrays within results_tuple has as many entries as available devices\n",
+    "    for i in xrange(len(available_devices)):\n",
+    "        # Collect the labels\n",
+    "        results_tuple[0].extend(sparse_tensor_value_to_texts(returns[0][i]))\n",
+    "        \n",
+    "        # Collect the decodings - at the moment we default to the first one \n",
+    "        results_tuple[1].extend(sparse_tensor_value_to_texts(returns[1][i][0]))\n",
+    "        \n",
+    "        # Collect the losses\n",
+    "        results_tuple[2].extend(returns[2][i])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's add a helper that bundles result collecting, WER calculation and WER report printing. It will return the mean WER."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def calculate_and_print_wer_report(caption, results_tuple):       \n",
+    "    # Calculating the individual WERs and the overall mean WER\n",
+    "    rates, wer = wers(results_tuple[0], results_tuple[1])\n",
+    "    \n",
+    "    # Printing the WER report\n",
+    "    print_wer_report(caption, wer, zip(results_tuple[0], results_tuple[1], rates, results_tuple[2]))\n",
+    "    \n",
+    "    # Return the mean WER\n",
+    "    return wer"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Finally we can assemble a helper routine that facilitates a full inference run plus report. It will return the mean WER."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "def run_inference(caption, data_set, results_params=None):\n",
+    "    if results_params is None:\n",
+    "        # Get the data_set specific graph end-points\n",
+    "        results_params = get_results_params(data_set)\n",
+    "        \n",
+    "    # Initialize the result collection tuple\n",
+    "    results = ([],[],[])\n",
+    "    \n",
+    "    # Loop over the batches\n",
+    "    for batch in range(int(ceil(float(data_set.total_batches) / len(available_devices)))):\n",
+    "        # Run and collect the results\n",
+    "        collect_results(results, session.run(results_params, feed_dict))\n",
```

Comment:
Yes, this is not very clean and also not following the overall approach. Fixed it.